src/sage/rings/polynomial/polynomial_element.pyx

```diff
@@ -2047,7 +2047,376 @@ cdef class Polynomial(CommutativePolynomial):
+
+        If we ask for an irreducible factor which does not exist, the function
+        will throw a ``ValueError``::
+
+            sage: # needs sage.rings.finite_rings
+            sage: F = GF(163)
+            sage: R.<x> = F[]
+            sage: g = (x - 1) * (x^3 + 7*x + 161)
+            sage: g.any_irreducible_factor(degree=2, assume_squarefree=True)
+            Traceback (most recent call last):
+            ...
+            ValueError: no irreducible factor of degree 2 could be computed from x^4 + 162*x^3 + 7*x^2 + 154*x + 2
+            sage: h = (x + 57) * (x + 98) * (x + 117) * (x + 145)
+
+        If we assume that the polynomial is product of irreducible polynomials of the
+        same degree, we must also supply the degree::
+
+            sage: # needs sage.rings.finite_rings
+            sage: F = GF(163)
+            sage: R.<x> = F[]
+            sage: h = (x + 57) * (x + 98) * (x + 117) * (x + 145)
+            sage: h.any_irreducible_factor(assume_distinct_deg=True)
+            Traceback (most recent call last):
+            ...
+            ValueError: degree must be known if distinct degree factorisation is assumed
+            sage: h.any_irreducible_factor(degree=1, assume_distinct_deg=True)   # random
+            x + 98
+
+        Also works for extension fields and even characteristic::
+
+            sage: F.<z4> = GF(2^4)
+            sage: R.<x> = F[]
+            sage: f = (x + z4^3 + z4^2)^4 * (x^2 + z4*x + z4) * (x^2 + (z4^3 + z4^2 + z4)*x + z4^2 + z4 + 1)
+            sage: f.any_irreducible_factor()
+            x + z4^3 + z4^2
+            sage: f.any_irreducible_factor(degree=2)  # random
+            x^2 + (z4^3 + z4^2 + z4)*x + z4^2 + z4 + 1
+
+        We can also use this function for polynomials which are not defined over finite
+        fields, but this simply falls back to a slow method of factorisation::
+
+            sage: R.<x> = ZZ[]
+            sage: f = 3*x^4 + 2*x^3
+            sage: f.any_irreducible_factor()
+            3*x + 2
+        """
+        # Make sure the user inputted something reasonable for degree
+        if degree is not None:
+            degree = ZZ(degree)
+            if degree < 1:
+                raise ValueError(f"{degree = } must be positive")
+
+        if assume_distinct_deg and degree is None:
+            raise ValueError("degree must be known if distinct degree factorisation is assumed")
+
+        # When not working over a finite field, do the simple thing of factoring.
+        # If degree is none, we return the first factor, otherwise we iterate
+        # through and look for a polynomial with the desired degree.
+        from sage.categories.finite_fields import FiniteFields
+        if self.base_ring() not in FiniteFields():
+            try:
+                factorisation = self.factor()
+            except (NotImplementedError, ValueError):
+                raise ValueError(f"Cannot factor {self} over the base ring {self.base_ring()}")
```

Comment:
```suggestion
                raise ValueError(f"cannot factor {self} over the base ring {self.base_ring()}")
```
All the standard exceptions in Python start with a lowercase letter; we should probably adhere to this convention.